### ISSUE TITLE
Automated Delete

### ISSUE THREAD

**Issue by denelon** [microsoft/winget-pkgs#22140]
# Description of the new feature/enhancement

Sometimes our automated validation isn't able to download an installer. When we run our periodic scan to test all manifests for obsolete (a.k.a. missing/can't download) packages, we get a false result.

# Proposed technical implementation details (optional)

**Ideas are welcome!**

One possibility is: apply some kind of waiver to indicate this is a known scenario so we don't try to automatically delete the affected installer once a waiver has been granted. We could then publish an issue on some recurring basis with the suspected installers. The community could validate and remove any of the installers / manifests for packages that are actually missing.

We would also need to clean up / remove the waivers when no manifests remain for a package.

**Comment by vedantmgoyal2009** [comment]
On my Azure VM, I ran my script for testing all URLs present in the repository, I was unable to download some links (even of GitHub). Since I was using Microsoft's Internet, I thought it is because of firewall settings applied at Microsoft's end. I guess the same is happening here.

**Comment by jedieaston** [comment]
Azure doesn't block sites across the entire cloud (well, at least not for this; there might be illegal stuff that is blocked for everything in the cloud). Although it's possible the winget team has the VMs behind an internal Microsoft firewall that blocks stuff I doubt that is what's happening. I don't think this is Microsoft's fault, I think that some sites might block or throttle traffic from Azure in case it's someone trying to do something nasty (I mean, who installs Calibre on a VM?).

My solution would be to install a VPN on the VM/container that checks the URLs so that the traffic doesn't appear to be coming from Azure as a test to see if we still have these issues or not. Apple uses a similar strategy for their app review internally (from what I've heard, I don't work there) so that apps don't see that traffic is coming from Apple and change their behavior. At a previous gig I had as a security intern, we had VPN accounts from a public provider so it didn't look like our traffic was coming from the corporate network for research. I'm sure someone in the Defender team at Microsoft has some solution for this that is battle-tested though :)

**Comment by vedantmgoyal2009** [comment]
> Although it's possible the winget team has the VMs behind an internal Microsoft firewall that blocks stuff I doubt that is what's happening.

I wanted to convey this only but I could not explain myself properly.

**Comment by donid** [comment]
[Here](https://stackoverflow.com/questions/68529103/why-does-webrequest-fail-with-404) is a report about similar problems with Urls taken from the winget package repo. The poster had an URL fail with 404 in c# as long as he didn't provide very specific request headers, but others reported that the URL didn't even work in a Browser (which always worked for the poster). Other Urls returned 403 when the UserAgent was missing. So, it seems quite hard to meet the requirements of some download servers...

**Comment by ItzLevvie** [comment]
CDNs, IP address and UAs (or User Agents) play a big role in this.

From my own testing using cURL which is inbox on Windows and was manually installed using apt-get on daily builds of Ubuntu 21.10: with CloudFlare DNS + most of the URLs were on Google Chrome Canary UA with some websites requiring custom UAs: I get a 99.96% pass coverage with 0.04% fail coverage. I haven't run into any sort of header issue yet. YMMW with PowerShell and with others.

Some websites will temporarily ban your IP from sending requests too quickly which is why I always run my script every 2 weeks or so; and some websites have CloudFlare enabled so when you download too quickly on a particular website, CloudFlare will think that you are a bot so you'll immediately be blocked.

**Comment by vedantmgoyal2009** [comment]
@ItzLevvie Which OS do you use for your testing?

**Comment by jedieaston** [comment]
I wonder if it would be worth using the GitHub mirrored URLs for Calibre instead of their hosted server (hosted on OVH, which has DDOS protection). We know GitHub doesn't screw with traffic from Azure.

**Comment by ItzLevvie** [comment]
> @ItzLevvie Which OS do you use for your testing?

Both Windows and Linux.

**Comment by jedieaston** [comment]
OT, but @ItzLevvie how long does it take you to download the entire repo?

**Comment by vedantmgoyal2009** [comment]
On Codespaces, I am getting 20 Gbps, is that because of region difference?

**Comment by vedantmgoyal2009** [comment]
> Oops, I mean different CAs in your Certificate Root Store between Windows and Linux. Windows seem to have the "SSL issuer certificates" issue fixed but Linux doesn't, so some of the 000 errors i.e. from `downloads.jtl-software.de` are related to that. Also my bad, the 000s for UAs in my previous reply was a mistake since it's been a long time I have had UA errors. UA errors are always in the 4XX range and the 000s come from SSL issuer certificates and timeouts (sometimes cURL error codes are confusing until you read the documentation).

Ok, so this means 000s are not UA errors but because of timeout and SSL.

> You're actually still seeing UA errors on Windows and Linux which can only be fixed until you pass the `--user-agent` parameter to cURL, but they're not in 000s but in the 4XX range. For example: from WhatsApp, from EVGA Precision X1 (removed from repository), from the `d1.music.126.net` domain, from the `ntwind.com` domain. Most of your 4XX errors are actually related to that because on my machine without any UA it fails and cURL ends up skipping the URL because it can't download because you passed an invalid UA. Without passing a valid UA, you will not know if the URL passes unless you tried it on a browser.

According to https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#406, UA error is 406 which I can see in `ntwind.com` and according to your reply, I can fix it with `curl --user-agent winget/1.0` as mentioned in https://github.com/microsoft/winget-pkgs/pull/24084#issuecomment-901281678

> I don't know what configuration you're running because I'm always on daily builds of Ubuntu with the latest dependencies (`sudo apt-get update && sudo apt-get upgrade`) so I'll always see a different result compared to you and you shouldn't be comparing my results with your results, it should just be a baseline for other people. My issues are only related to the "SSL issuer certificates" (on cURL + Linux) and nothing else. Windows is fine.

I am on [these](http://cdimage.ubuntu.com/daily-live/pending/) builds because it is my habit 😀 to sign up for beta/nightly/insider/etc... everywhere.

If I am wrong anywhere

**Comment by ItzLevvie** [comment]
> OT, but @ItzLevvie how long does it take you to download the entire repo?

For downloading all of the URLs in the repository, it takes a couple of hours with no timeouts or anything. It's a bit more than what it takes to download an AOSP repository with `--depth 1` because of CDN (i.e. some of the Chinese URLs take a bit long to download because of how far I am between where I live and China).

>... I don't get SSL errors but the timeout and UA errors only.

For Windows, sure. You're not getting any SSL errors on Windows because of the differences between WinSSL v.s. OpenSSL.

but for Linux? No, you're still getting those errors on Linux. Just like how timeout errors are 000, SSL issuer errors are also 000. 

You've to check all of the 000 URLs without the timeout and UA to check why it's erroring out and fix them in your script.

Also, when cURL returns 301 or 302 for the URLs you've hidden from the output, it's not actually returning the proper error code. It's still 404 Not Found or 403 Forbidden even if the URL is 301 or 302 (for 200, it's fine) and using Windows won't fix that either. That's just the way that the developer designed their website to work.

Pro Tip: You can use the RegEx functionality in Notepad++ which should only take less than 30 seconds to do and filter out 200, 302 (from `https://github.com/) and test rest of the URLs that weren't filtered out. Rather than filtering all 301 & 302 error codes in your script.

>According to https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#406, UA error is 406 which I can see in ntwind.com and according to your reply, I can fix it with curl --user-agent winget/1.0 as mentioned in #24084 (comment)

UA errors also happen on the 400 and 403 error codes not just 406.

Your output from GitHub Actions shows that you'll still get those 000 errors even without the timeout and user agent. Rerun that script without any timeouts and UA and then check cURL's output log.

>What I have seen is curl of Windows is just an alias of Invoke-WebRequest

This means you weren't using the built-in cURL which was introduced in 17763 and above. Invoke-WebRequest doesn't launch `C:\Windows\System32\curl.exe`.

You can use cURL from Command Prompt (i.e. `curl --version`), or use cURL in Windows PowerShell (i.e. `C:\Windows\System32\curl.exe --version`).

**Comment by vedantmgoyal2009** [comment]
You tested 4000+ URLs manually, which return 302 HTTP codes? 🤯

**Comment by ItzLevvie** [comment]
>Did you test 4000+ URLs manually, which return 302 HTTP codes? 🤯

No, there are roughly less than 100 (302/301) URLs left for me when properly filtered out and it takes less than 5 minutes to go through all of them to manually check if it's still working. Your results will most likely be different from mine.

>Also, can I use UA "winget/1.0" because we have to test whether the URL is accessible by winget and not test whether it is accessible by browser?

Sure, that works too and is better than using the browser UA because GPU-Z now needs the "winget/1.0" user agent to pass.

>Will https://major.io/2007/03/23/exporting-ssl-certificates-from-windows-to-linux/ help me with SSL errors?

No idea.

**Comment by vedantmgoyal2009** [comment]
> > Will https://major.io/2007/03/23/exporting-ssl-certificates-from-windows-to-linux/ help me with SSL errors?
> 
> No idea.

Thanks for helping me. I will test this if it works.
